core/arch/arm/mm/core_mmu.c

```diff
@@ -478,66 +544,78 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 			map->region_size = SMALL_PAGE_SIZE;
 		else
 			panic("Impossible memory alignment");
+
+#ifdef CFG_WITH_PAGER
+		if (map_is_tee_ram(map))
+			map->region_size = SMALL_PAGE_SIZE;
+#endif
 	}
 
 	/*
-	 * bootcfg_memory_map is sorted in order first by type and last by
-	 * address. This puts TEE_RAM first and TA_RAM second
-	 *
+	 * To ease mapping and lower use of xlat tables, sort mapping
+	 * description moving small-page regions after the pgdir regions.
 	 */
-	map = memory_map;
-	assert(map->type == MEM_AREA_TEE_RAM);
-	map->va = map->pa;
-#ifdef CFG_WITH_PAGER
-	map->region_size = SMALL_PAGE_SIZE;
+	qsort(memory_map, last, sizeof(struct tee_mmap_region),
+		cmp_mmap_by_bigger_region_size);
+
+#if !defined(CFG_WITH_LPAE)
+	/*
+	 * 32bit MMU descriptors cannot mix secure and non-secure mapping in
+	 * the same level2 table. Hence sort secure mapping from non-secure
+	 * mapping.
+	 */
+	for (count = 0, map = memory_map; map_is_pgdir(map); count++, map++)
+		;
+
+	qsort(memory_map + count, last - count, sizeof(struct tee_mmap_region),
+		cmp_mmap_by_secure_attr);
 #endif
-	map->attr = core_mmu_type_to_attr(map->type);
 
+	/* Map flat mapped addresses first */
+	va = (vaddr_t)~0UL;
+	flat_map_va_max = 0;
+	for (map = memory_map; map->type != MEM_AREA_NOTYPE; map++) {
+		if (!map_is_flat_mapped(map))
+			continue;
+
+		map->attr = core_mmu_type_to_attr(map->type);
+		map->va = map->pa;
+		va = MIN(va, ROUNDDOWN(map->va, map->region_size));
+		flat_map_va_max = MAX(va, ROUNDUP(flat_map_va_max, map->size));
+
+	}
+	assert(va >= CFG_TEE_RAM_START);
+	assert(flat_map_va_max <= CFG_TEE_RAM_START + CFG_TEE_RAM_VA_SIZE);
+
+	if (core_mmu_place_tee_ram_at_top(va)) {
```

Comment:
It wouldn't hurt with a comment stating that `va` is the lowest flat mapped address, even if it's obvious for us two for the moment.